1. (1 балл) Загрузите предложенный вам датасет с помощью функции `sklearn.datasets.fetch_openml`. Выведите текстовое описание загруженного датасета. Обозначьте целевую переменную за `y`, а остальные данные за `X`.

In [1]:
from sklearn.datasets import fetch_openml

In [2]:
df = fetch_openml(name="JuanFeldmanIris")
df

{'data':      sepallength  sepalwidth  petallength  petalwidth
 0            5.1         3.5          1.4         0.2
 1            4.9         3.0          1.4         0.2
 2            4.7         3.2          1.3         0.2
 3            4.6         3.1          1.5         0.2
 4            5.0         3.6          1.4         0.2
 ..           ...         ...          ...         ...
 145          6.7         3.0          5.2         2.3
 146          6.3         2.5          5.0         1.9
 147          6.5         3.0          5.2         2.0
 148          6.2         3.4          5.4         2.3
 149          5.9         3.0          5.1         1.8
 
 [150 rows x 4 columns],
 'target': 0         Iris-setosa
 1         Iris-setosa
 2         Iris-setosa
 3         Iris-setosa
 4         Iris-setosa
             ...      
 145    Iris-virginica
 146    Iris-virginica
 147    Iris-virginica
 148    Iris-virginica
 149    Iris-virginica
 Name: class, Length: 150, dtype: category

In [3]:
df.DESCR # Текстовое описание датасета

'Iris DataSet\n\nDownloaded from openml.org.'

In [4]:
y = df.target # Целевая переменная
y

0         Iris-setosa
1         Iris-setosa
2         Iris-setosa
3         Iris-setosa
4         Iris-setosa
            ...      
145    Iris-virginica
146    Iris-virginica
147    Iris-virginica
148    Iris-virginica
149    Iris-virginica
Name: class, Length: 150, dtype: category
Categories (3, object): ['Iris-setosa', 'Iris-versicolor', 'Iris-virginica']

In [5]:
X = df.data # Сами данные
X

sepallength  sepalwidth  petallength  petalwidth
0            5.1         3.5          1.4         0.2
1            4.9         3.0          1.4         0.2
2            4.7         3.2          1.3         0.2
3            4.6         3.1          1.5         0.2
4            5.0         3.6          1.4         0.2
..           ...         ...          ...         ...
145          6.7         3.0          5.2         2.3
146          6.3         2.5          5.0         1.9
147          6.5         3.0          5.2         2.0
148          6.2         3.4          5.4         2.3
149          5.9         3.0          5.1         1.8

[150 rows x 4 columns]

2. (1 балл) Выведите основную статистическую информацию о данных. Сделайте количественное описание датасета: число строк (объектов), число столбцов (признаков), статистику по признакам, количество классов (значений целевой переменной).

In [6]:
X.describe()

sepallength  sepalwidth  petallength  petalwidth
count   150.000000  150.000000   150.000000  150.000000
mean      5.843333    3.054000     3.758667    1.198667
std       0.828066    0.433594     1.764420    0.763161
min       4.300000    2.000000     1.000000    0.100000
25%       5.100000    2.800000     1.600000    0.300000
50%       5.800000    3.000000     4.350000    1.300000
75%       6.400000    3.300000     5.100000    1.800000
max       7.900000    4.400000     6.900000    2.500000

In [7]:
y.describe()

count             150
unique              3
top       Iris-setosa
freq               50
Name: class, dtype: object

In [8]:
X.shape, y.shape

((150, 4), (150,))

Количество строк и столбцов: 4 столбца с признаками, 1 столбец в целевой переменной, 150 строк и в самих данных, и в целевой переменной.

3. (1 балл) Убедитесь, что данные пригодны для моделирования. В данных не должно быть пропущенных значений, все признаки должны быть численными. Если эти условия нарушаются, исправьте это.

In [9]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   sepallength  150 non-null    float64
 1   sepalwidth   150 non-null    float64
 2   petallength  150 non-null    float64
 3   petalwidth   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


In [10]:
y.info()

<class 'pandas.core.series.Series'>
RangeIndex: 150 entries, 0 to 149
Series name: class
Non-Null Count  Dtype   
--------------  -----   
150 non-null    category
dtypes: category(1)
memory usage: 414.0 bytes


In [11]:
X.isna().sum()

sepallength    0
sepalwidth     0
petallength    0
petalwidth     0
dtype: int64

In [12]:
y.isna().sum()

0

Пропущенных значений нет, все признаки являются численными

4. (2 балла) Обучите модель логистической регрессии на рассматриваемых данных из библиотеки sklearn. Рассчитайте метрики accuracy и выведите таблицу классификации.

In [13]:
# Перед работой нормализируем данные.
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X = scaler.fit_transform(X)

In [14]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
import time

In [15]:
# Обучение модели
start_time = time.time()
log_reg = LogisticRegression(max_iter=10000)
log_reg.fit(X, y)
train_time_log_reg = time.time() - start_time

In [16]:
# Оценка модели
y_pred = log_reg.predict(X)
accuracy_log_reg = accuracy_score(y, y_pred)
print(f"Accuracy (Logistic Regression): {accuracy_log_reg}")
print(f"Время обучения: {train_time_log_reg:.4f} секунд")
print("Classification Report:\n", classification_report(y, y_pred, zero_division=0))
print("Confusion Matrix:\n", confusion_matrix(y, y_pred))

Accuracy (Logistic Regression): 0.9733333333333334
Время обучения: 0.0833 секунд
Classification Report:
                  precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        50
Iris-versicolor       0.98      0.94      0.96        50
 Iris-virginica       0.94      0.98      0.96        50

       accuracy                           0.97       150
      macro avg       0.97      0.97      0.97       150
   weighted avg       0.97      0.97      0.97       150

Confusion Matrix:
 [[50  0  0]
 [ 0 47  3]
 [ 0  1 49]]


Метрика accuracy:
Общая точность модели составляет 0.97. Высокое значение точности указывает на то, что модель хорошо справляется с задачей классификации в текущем состоянии.

Метрика macro avg:
Средние значения для всех классов 0.97, что указывает на высокую общую производительность.

Метрика weighted avg:
Учитывая сбалансированность классов, взвешенная средняя точность равна 0.97, что подтверждает высокую производительность.

Confusion Matrix:
Матрица показывает, что модель делает малое количество ошибок.

Classification Report:
Precision, Recall, и F1-Score не сильно варьируются между классами (метрики высоки, значения близки к 1).

5. (2 балла) Обучите полиномиальную модель классификации. Рассчитайте метрики accuracy и выведите таблицу классификации. Попробуйте разные степени полинома и выберите ту, которая работает лучше.

In [17]:
from sklearn.preprocessing import PolynomialFeatures
import time

In [18]:
degrees = [1, 2, 3, 4, 5]  # Пробуем разные степени полинома
best_degree = None
best_accuracy = 0
results_poly = []

for degree in degrees:
    # Преобразование данных с помощью PolynomialFeatures
    poly = PolynomialFeatures(degree=degree)
    X_poly = poly.fit_transform(X)
    
    # Создание и обучение модели логистической регрессии
    model = LogisticRegression(max_iter=1000)
    start_time = time.time()
    model.fit(X_poly, y)
    end_time = time.time()
    train_time = end_time - start_time
    results_poly.append(train_time)
    
    # Оценка точности
    y_pred = model.predict(X_poly)
    accuracy = accuracy_score(y, y_pred)
    
    print(f"Полином степени {degree}")
    print(f"Accuracy: {accuracy:.4f}")
    print(f"Время обучения: {train_time:.4f} секунд")
    print("Classification Report:")
    print(classification_report(y, y_pred, zero_division=0))
    print("Confusion Matrix:")
    print(confusion_matrix(y, y_pred))
    print("-" * 50)
    
    # Сохраняем лучшую степень
    if accuracy > best_accuracy:
        best_accuracy = accuracy
        best_degree = degree

# Итог
print(f"Лучшая степень полинома: {best_degree} с accuracy {best_accuracy:.4f}")

Полином степени 1
Accuracy: 0.9733
Время обучения: 0.0040 секунд
Classification Report:
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        50
Iris-versicolor       0.98      0.94      0.96        50
 Iris-virginica       0.94      0.98      0.96        50

       accuracy                           0.97       150
      macro avg       0.97      0.97      0.97       150
   weighted avg       0.97      0.97      0.97       150

Confusion Matrix:
[[50  0  0]
 [ 0 47  3]
 [ 0  1 49]]
--------------------------------------------------
Полином степени 2
Accuracy: 0.9867
Время обучения: 0.0040 секунд
Classification Report:
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        50
Iris-versicolor       0.98      0.98      0.98        50
 Iris-virginica       0.98      0.98      0.98        50

       accuracy                           0.99       150
      macro avg       0.99   

Итак, лучшие результаты показывает модель со степенью полинома 2 (да, аналогичные результаты показывают и модели с полиномами 3, 4, 5).

6. (2 балла) Обучите модель классификации по методу опорных векторов. Рассчитайте метрики accuracy и выведите таблицу классификации. Попробуйте разные ядерные функции и выберите ту, которая работает лучше.

In [19]:
from sklearn.svm import SVC

In [20]:
kernels = ['linear', 'poly', 'rbf', 'sigmoid']
best_kernel = None
best_accuracy_svc = 0
results = []

for kernel in kernels:
    start_time = time.time()
    svm = SVC(kernel=kernel)
    svm.fit(X, y)
    train_time = time.time() - start_time

    y_pred = svm.predict(X)
    accuracy = accuracy_score(y, y_pred)
    results.append((kernel, accuracy, train_time))

    if accuracy > best_accuracy_svc:
        best_kernel = kernel
        best_accuracy_svc = accuracy

    print(f"Kernel: {kernel}, Accuracy: {accuracy}, Time: {train_time:.2f}s")

print(f"\nBest Kernel: {best_kernel}, Best Accuracy: {best_accuracy_svc}")

Kernel: linear, Accuracy: 0.9666666666666667, Time: 0.00s
Kernel: poly, Accuracy: 0.9533333333333334, Time: 0.00s
Kernel: rbf, Accuracy: 0.9733333333333334, Time: 0.00s
Kernel: sigmoid, Accuracy: 0.8866666666666667, Time: 0.00s

Best Kernel: rbf, Best Accuracy: 0.9733333333333334


Итак, лучшие результаты показывает модель с ядром rbf (общая точность модели примерно 0.97)

7. (1 балл) Обучите модель классификации Перцептрон. Рассчитайте метрики accuracy и выведите таблицу классификации. 

In [21]:
from sklearn.linear_model import Perceptron

In [22]:
start_time = time.time()
perceptron = Perceptron(max_iter=1000)
perceptron.fit(X, y)
train_time_perc = time.time() - start_time

y_pred = perceptron.predict(X)
accuracy = accuracy_score(y, y_pred)

print(f"Accuracy (Perceptron): {accuracy}")
print(f"Время обучения: {train_time_perc:.4f} секунд")
print("Classification Report:\n", classification_report(y, y_pred, zero_division=0))
print("Confusion Matrix:\n", confusion_matrix(y, y_pred))

Accuracy (Perceptron): 0.8666666666666667
Время обучения: 0.0016 секунд
Classification Report:
                  precision    recall  f1-score   support

    Iris-setosa       0.96      1.00      0.98        50
Iris-versicolor       1.00      0.60      0.75        50
 Iris-virginica       0.74      1.00      0.85        50

       accuracy                           0.87       150
      macro avg       0.90      0.87      0.86       150
   weighted avg       0.90      0.87      0.86       150

Confusion Matrix:
 [[50  0  0]
 [ 2 30 18]
 [ 0  0 50]]


Общая точность модели составляет 0.87. Весьма высокое значение точности указывает на то, что модель хорошо справляется с задачей классификации в текущем состоянии.

Метрика macro avg: Средние значения не меньше 0.86, что указывает на высокую общую производительность.

Метрика weighted avg: Учитывая сбалансированность классов, взвешенная средняя точность не меньше 0.86, что подтверждает высокую производительность.

Confusion Matrix: Матрица показывает, что модель всё же далека от идеальной.

Classification Report: Precision, Recall, и F1-Score средне варьируются между классами.

8. (2 балла) Выведите итоговую таблицу сравнения всех моделей. В таблице должна быть информация о эффективности и времени обучения каждой модели. Сделайте вывод о применимости и эффективности моделей для классификации объектов в данной задаче.

In [24]:
import pandas as pd

In [25]:
comparison = pd.DataFrame({
    "Model": ["Logistic Regression", "Polynomial Logistic Regression", 
              "SVM (Best Kernel)", "Perceptron"],
    "Accuracy": [accuracy_log_reg, 
                 best_accuracy,
                 best_accuracy_svc,
                 accuracy],
    "Training Time (s)": [train_time_log_reg, 
                          results_poly[best_degree-1], 
                          results[kernels.index(best_kernel)][2],
                          train_time_perc ]
})

print(comparison)

                            Model  Accuracy  Training Time (s)
0             Logistic Regression  0.973333           0.083279
1  Polynomial Logistic Regression  0.986667           0.004000
2               SVM (Best Kernel)  0.973333           0.000000
3                      Perceptron  0.866667           0.001630


### Заключительный вывод

На основе предоставленных данных сравнения моделей классификации:

1. **Логистическая регрессия (Logistic Regression):**
   - **Точность (Accuracy):** 0.97
   - **Время обучения:** 0.08 секунды
   - Быстро обучается, хорошо справляется с задачей классификации. В нашем контексте является самой худшей моделью по скорости обучения, однако показатели времени всё же малы.

2. **Полиномиальная логистическая регрессия (Polynomial Logistic Regression):**
   - **Точность (Accuracy):** 0.987
   - **Время обучения:**  0.004 секунд
   - Отличная точность, высокая скорость обучения. Подходит для задач с нелинейной зависимостью между признаками, но может требовать значительных ресурсов, особенно при высокой степени полинома (однако не в нашем задаче).

3. **Метод опорных векторов (SVM с лучшим ядром):**
   - **Точность (Accuracy):** 0.97
   - **Время обучения:** 0.000... секунды (крайне мало)
   - Отличная точность и быстрое обучение. В общем и целом хорошо подходит для задач, где данные имеют более сложную структуру, но при этом объем вычислений остается низким.

4. **Перцептрон (Perceptron):**
   - **Точность (Accuracy):** 0.866
   - **Время обучения:** 0.0016 секунды
   - Худшая модель по точности в нашей задаче. Неплохо справляется с задачей классификации в текущем контексте, однако у нас имеются показатели лучше ( Polynomial Logistic Regression с accuracy около 0.987, например).
---

### Итог:
Можно утверждать, что для нашей задачи все используемые методы дают положительный результат (разве что у Perceptron показатели ниже остальных). Однако для данной задачи наилучший выбор — **полиномиальная логистическая регрессия**, так как она обеспечивает наивысшую точность (accuracy = 0.986667). Время обучения во всех случаях довольно мало.

### Примечание:
Значения accuracy и training time для подведения итогов взяты с примерным значащим количеством знаков после запятой.